src/reader.cpp

```diff
@@ -1206,6 +1233,68 @@ void reader_data_t::maybe_refilter_pager(const editable_line_t *el) {
+                                                   history_search_direction_t direction,
+                                                   size_t history_index,
+                                                   const wcstring &search_string) {
+    constexpr size_t history_pager_max_size = 12;
+    completion_list_t completions;
+    history_search_t search{history, search_string, history_search_type_t::contains,
+                            smartcase_flags(search_string), history_index};
+    bool can_go_back = search.go_to_next_match(~direction, true /* probe_only */);
+    while (completions.size() < history_pager_max_size && search.go_to_next_match(direction)) {
+        const history_item_t &item = search.current_item();
+        completions.push_back(completion_t{
+            item.str(), L"", string_fuzzy_match_t::exact_match(),
+            COMPLETE_REPLACES_COMMANDLINE | COMPLETE_DONT_ESCAPE | COMPLETE_DONT_SORT});
+    }
+    size_t last_index = search.current_index();
+    if (direction == history_search_direction_t::forward)
+        std::reverse(completions.begin(), completions.end());
+    bool can_go_further = search.go_to_next_match(direction, true /* probe_only */);
+    return {completions, last_index, can_go_back, can_go_further};
+}
+
+void reader_data_t::fill_history_pager(bool new_search, history_search_direction_t direction) {
+    assert(!new_search || direction == history_search_direction_t::backward);
+    size_t index;
+    if (new_search) {
+        index = 0;
+    } else if (direction == history_search_direction_t::forward) {
+        index = history_pager_history_index_start;
+    } else {
+        assert(direction == history_search_direction_t::backward);
+        index = history_pager_history_index_end;
+    }
+    const wcstring &search_term = pager.search_field_line.text();
+    auto shared_this = this->shared_from_this();
+    debounce_history_pager().perform(
+        [=]() { return history_pager_search(shared_this->history, direction, index, search_term); },
+        [=](const history_pager_result_t &result) {
+            if (search_term != shared_this->pager.search_field_line.text())
+                return;  // Stale request.
+            if (result.matched_commands.empty() && !new_search) {
+                // No more matches, keep the existing ones and flash.
+                shared_this->flash();
+                return;
+            }
+            if (direction == history_search_direction_t::forward) {
+                shared_this->history_pager_history_index_start = result.final_index;
+                shared_this->history_pager_history_index_end = index;
+            } else {
+                shared_this->history_pager_history_index_start = index;
+                shared_this->history_pager_history_index_end = result.final_index;
+            }
+            shared_this->pager.extra_progress_text = (result.can_go_back || result.can_go_further)
```

Comment:
This should use the direction - `((direction == history_search_direction_t::forward && result.can_go_further) || (direction == history_search_direction_t::backward && result.can_go_back))`

Otherwise you'll get it if there is more than one page.

Reply:
Ok, so we don't need to indicate if the opposite direction is available?

Reply:
Yeah that's probably the simplest thing to do for now. In future we can also display the index of the match, and maybe the timestamp (or should that go into the description)

Reply:
>In future we can also display the index of the match

Do you mean the index as in "number of history entry"? I don't think that helps much.

Page X/Y would be ideal, but of course that requires figuring out how many pages there are.

>Ok, so we don't need to indicate if the opposite direction is available?

Yeah my idea was that the opposite direction is, after one page, always available.

Of course that doesn't tell people that going back is a thing they can generally do.

>maybe the timestamp (or should that go into the description)

Timestamp descriptions are an interesting idea. They would solve the issue of seeing where entries end.

But it might be overkill, especially if you do full "YYYY-MM-DD HH:MM:SS" everywhere?

Reply:
Oooh, how about "Page 4 (search again for more)"?